File: shared/src/main/java/io/github/robertomike/hefesto/utils/CastUtils.java
package io.github.robertomike.hefesto.utils;

import io.github.robertomike.hefesto.exceptions.HefestoException;
import io.github.robertomike.hefesto.exceptions.QueryException;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.math.BigInteger;

/**
 * This class is a helper to cast values
 */
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class CastUtils {
    /**
     * Casts a value to the specified type.
     *
     * @param  type   the target class type for the casting
     * @param  value  the value to be casted
     * @param <T>     the type of the value
     * @return        the casted value of the specified type
     */
    public static <T> Object castValue(Class<T> type, Object value) {
        if (type.isInstance(value)) {
            return value;
        }

        if (type.equals(String.class)) {
            return type.toString();
        }
        if (type.equals(Integer.class) || type.equals(int.class)) {
            return Integer.valueOf(value.toString());
        }
        if (type.equals(Byte.class) || type.equals(byte.class)) {
            return Integer.valueOf(value.toString());
        }
        if (type.equals(Double.class) || type.equals(double.class)) {
            return Double.valueOf(value.toString());
        }
        if (type.equals(Short.class) || type.equals(short.class)) {
            return Short.valueOf(value.toString());
        }
        if (type.equals(Long.class) || type.equals(long.class)) {
            return Long.valueOf(value.toString());
        }
        if (type.equals(BigDecimal.class)) {
            return BigDecimal.valueOf(Double.parseDouble(value.toString()));
        }
        if (type.equals(BigInteger.class)) {
            return BigInteger.valueOf(Long.parseLong(value.toString()));
        }
        if (type.equals(Boolean.class) || type.equals(boolean.class)) {
            return Boolean.valueOf(value.toString());
        }
        if (type.equals(char.class)) {
            return value.toString().charAt(0);
        }

        throw new QueryException("Unsupported casting type:" + type);
    }

    public static <T> T getClassInstance(Class<T> clazz) {
        try {
            return clazz.getConstructor().newInstance();
        } catch (Exception e) {
            throw new HefestoException("Error creating class instance for " + clazz.getName() , e);
        }
    }
}

File: shared/src/main/kotlin/io/github/robertomike/hefesto/utils/CastUtils.kt
package io.github.robertomike.hefesto.utils

import io.github.robertomike.hefesto.exceptions.HefestoException
import io.github.robertomike.hefesto.exceptions.QueryException
import java.math.BigDecimal
import java.math.BigInteger

/**
 * Utility object for type casting and class instantiation operations.
 * Provides safe casting methods for common data types and dynamic object creation.
 * 
 * This utility handles conversions between various numeric types, strings, booleans,
 * and other common Java types used in query results.
 */
object CastUtils {
    /**
     * Casts a value to the specified type.
     *
     * @param  type   the target class type for the casting
     * @param  value  the value to be casted
     * @param <T>     the type of the value
     * @return        the casted value of the specified type
     */
    @JvmStatic
    fun <T> castValue(type: Class<T>, value: Any): Any {
        if (type.isInstance(value)) {
            return value
        }

        if (type == String::class.java || type == java.lang.String::class.java) {
            return value.toString()
        }
        if (type == Int::class.javaPrimitiveType || type == Integer::class.java || type == java.lang.Integer::class.java) {
            return value.toString().toInt()
        }
        if (type == Byte::class.javaPrimitiveType || type == java.lang.Byte::class.java) {
            return value.toString().toByte()
        }
        if (type == Double::class.javaPrimitiveType || type == java.lang.Double::class.java) {
            return value.toString().toDouble()
        }
        if (type == Short::class.javaPrimitiveType || type == java.lang.Short::class.java) {
            return value.toString().toShort()
        }
        if (type == Long::class.javaPrimitiveType || type == java.lang.Long::class.java) {
            return value.toString().toLong()
        }
        if (type == BigDecimal::class.java) {
            return BigDecimal.valueOf(value.toString().toDouble())
        }
        if (type == BigInteger::class.java) {
            return BigInteger.valueOf(value.toString().toLong())
        }
        if (type == Boolean::class.javaPrimitiveType || type == java.lang.Boolean::class.java) {
            return value.toString().toBoolean()
        }
        if (type == Char::class.javaPrimitiveType || type == Character::class.java) {
            return value.toString()[0]
        }

        throw QueryException("Unsupported casting type: $type")
    }

    /**
     * Creates a new instance of the specified class using its no-argument constructor.
     * 
     * This method is typically used internally for DTO instantiation during query result mapping.
     *
     * @param clazz the class to instantiate
     * @param <T> the type of the class
     * @return a new instance of the specified class
     * @throws HefestoException if the class cannot be instantiated (e.g., no default constructor)
     */
    @JvmStatic
    fun <T> getClassInstance(clazz: Class<T>): T {
        return try {
            clazz.getConstructor().newInstance()
        } catch (e: Exception) {
            throw HefestoException("Error creating class instance for ${clazz.name}", e)
        }
    }
}
